### DISCUSSION TITLE
Storage API definition

### DISCUSSION THREAD

**Discussion by fraimondo** [juaml/junifer#87] [Ideas]
In order to properly defined the storage API (the interface between the markers output and the storage input), we need to know which objects are likely to be produced by the markers.

For example, a functional connectivity marker will output a 2D matrix (maybe triangular) in which the rows and columns have names.  A sphere aggregator (or parcel aggregator) will output a vector in which all the columns have names. 

So the question is, which other type do you envision a marker will likely output.

Keep in mind that it should be a general marker and not a highly specific use-case for which Junifer will not be applicable.

So far, we have:

* vector with named elements (columns)
* 2d matrix with named elements (columns and rows)
* timeseries with named elements (columns only, timepoints are indexes)
* nifti images (the marker will output a nifti image directly)

**Comment by fraimondo** [comment]
@kaurao @samihamdan @LeSasse @antogeo @juaml/hackers @synchon @omidvarnia 

Please discuss here

**Reply by LeSasse** [reply]
I cannot think of a different type at the moment, but one thing to keep in mind if you want to implement lower triangular FC, it will be good to keep it consistent with the corresponding [nilearn function](https://nilearn.github.io/dev/modules/generated/nilearn.connectome.sym_matrix_to_vec.html) that I am sure you are aware of, just wanted to explicitly record that here.

**Reply by fraimondo** [reply]
We have support for lower and upper triangular matrixes, including or excluding the diagonal

**Reply by omidvarnia** [reply]
I also believe that your list of output types is pretty comprehensive (for my work) and covers the necessary formats.

**Reply by LeSasse** [reply]
> We have support for lower and upper triangular matrixes, including or excluding the diagonal

thats good, but in that case it will be also quite nice to 'hint' in the documentation to beginner users which functions they should use to correctly reconstruct the full matrix if they want to (I know it seems trivial but its an easy mistake and can end up costing time and nerves.)

**Reply by fraimondo** [reply]
That's a good one. We'll keep that in mind for the reader.

**Comment by fraimondo** [comment]
This was already addressed in #103.
